### PR TITLE
Add Override for data Product path so user can define whatever path they like

### PR DESCRIPTION
**Pull request by jchate6**
This should allow a user to add a `DATA_PRODUCT_PATH` value to their `settings.py` allowing for them to override the default path creation method for Data Products.

### CODE REVIEW

File: docs/managing_data/customizing_data_processing.rst [hunk 1]
```diff
@@ -48,6 +48,21 @@ In order to add new data product types, simply add a new key/value pair,
...
+When a ``DataProduct`` is uploaded, its path is set by a `data_product_path`. By default this is set to
+``{target}/{facility}/{filename}``, but can be overridden by setting the ``DATA_PRODUCT_PATH`` in
+``settings.py`` to a dot-separated method path that takes a ``DataProduct`` and filename as arguments. For example, if
+you wanted to set the `data_product_path` to ``{target}/{facility}/{observation_id}/{filename}``, you would point the
+``DATA_PRODUCT_PATH`` to a custom method that looks something like this:
+
+.. code:: python
+
+   def custom_data_product_path(data_product, filename):
+       return '{}/{}/{}/{}'.format(data_product.target.name,
```
Comment (phycodurus) [inline docs/managing_data/customizing_data_processing.rst:60]: Let's use an f-string here.

File: tom_dataproducts/models.py [hunk 2]
```diff
@@ -85,11 +93,21 @@ def data_product_path(instance, filename):
...
-    else:
-        return '{0}/none/{1}'.format(instance.target.name, filename)
+    try:
+        path_class = settings.DATA_PRODUCT_PATH
+        try:
+            mod_name, class_name = path_class.rsplit('.', 1)
+            mod = import_module(mod_name)
+            clazz = getattr(mod, class_name)
+        except (ImportError, AttributeError):
+            raise ImportError('Could not import {}. Did you provide the correct path?'.format(path_class))
```
Comment (phycodurus) [inline tom_dataproducts/models.py:103]: let's use an f-string here as well.

File: tom_dataproducts/models.py [hunk 3]
```diff
@@ -85,11 +93,21 @@ def data_product_path(instance, filename):
...
+        return clazz(instance, filename)
+    except AttributeError:
+        # Uploads go to MEDIA_ROOT
+        if instance.observation_record is not None:
+            return '{0}/{1}/{2}'.format(instance.target.name, instance.observation_record.facility, filename)
+        else:
+            return '{0}/none/{1}'.format(instance.target.name, filename)
```
Comment (phycodurus) [inline tom_dataproducts/models.py:110]: more f-string opportunities :-)

**Review by phycodurus — APPROVED** [review]

**Comment by davner** [comment]
This looks extremely promising, thank you! I can get around to testing it later today or first thing tomorrow.

File: docs/managing_data/customizing_data_processing.rst (hunk 1 above)
Comment (jchate6) [reply docs/managing_data/customizing_data_processing.rst:60]: the line breaks are so much uglier... :(

**Pull request merged**